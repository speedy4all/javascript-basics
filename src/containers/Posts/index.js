import React, { useCallback, useEffect, useMemo, useState } from "react";
import  { connect } from 'react-redux'

import PostsView from "../../components/PostsView";
import SearchView from "../../components/SearchView";

import { setPosts, setLoadingFlag } from '../../actions/posts';
import { setError } from '../../actions/error';

let Posts = (props) => {
  const { posts, error, isLoading } = props;
  const [searchParam, setSearchParam] = useState("");

  const fetchPosts = () => {
    props.dispatchSetLoading(true);
    fetch("https://jsonplaceholder.typicode.com/posts")
        .then((response) => response.json())
        .then((data) => {
          props.dispatchSetPosts(data)
          props.dispatchSetLoading(false);

          // simulare eroare
          // setTimeout(() => {
          //   props.dispatchSetError({
          //     code: 400,
          //     message: 'Eroare 400'
          //   })
          // }, 5000)

        })
  };

  useEffect(() => {

    if (!searchParam || !isLoading) {
      fetchPosts();
    }
  }, [searchParam]);

  const filteredPosts = useMemo(() => {
    if (!searchParam) {
      return posts;
    }
    return posts.filter(({ body }) => body.includes(searchParam));
  }, [searchParam, posts, !isLoading]);

  let renderPosts = <PostsView posts={filteredPosts} />;
  if(isLoading) {
    renderPosts = <div>Pagina se incarca!</div>
  }
  if(error?.message) {
    renderPosts = <div>{error.message}</div>
  }

  return (
    <>
      <SearchView search={searchParam} setSearch={setSearchParam} />
      {renderPosts}
    </>
  );
}

Posts.propTypes = {};

const mapStateToProps = (state, ownProps) => {
  // "state" - este state-ul aplicatiei (redux)
  // ownProps - este props-urile componentei
  // console.log({state, ownProps});
  return {
    posts: state.posts.list,
    error: state.error,
    isLoading: state.posts.loading,
  }
}

const mapDispatchToProps = {
  dispatchSetPosts: setPosts,
  dispatchSetError: setError,
  dispatchSetLoading: setLoadingFlag,
}

Posts = connect(mapStateToProps, mapDispatchToProps)(Posts)

export default Posts



