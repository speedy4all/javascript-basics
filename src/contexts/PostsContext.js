import React, { useState, useContext, useMemo }  from 'react'

export const PostsContext = React.createContext();

export function usePosts() {
    return useContext(PostsContext)
}

export function PostProvider({ children}) {
    const [posts, setPosts ] = useState([]);
    const [searchParam, setSearchParam] = useState("");

    const fetchPosts = () =>
        fetch("https://jsonplaceholder.typicode.com/posts")
            .then((response) => response.json())
            .then((data) => setPosts(data));

    const deletePost = (id) => {
    }

    const filteredPosts = useMemo(() => {
      if (!searchParam) {
        return posts;
      }
      return posts.filter(({ body }) => body.includes(searchParam));
    }, [searchParam, posts]);

    const value = {
        posts,
        filteredPosts,
        searchParam,
        fetchPosts,
        setSearchParam,
        deletePost,
    }

    return (
        <PostsContext.Provider value={value}>
            {children}
        </PostsContext.Provider>
    )
}
