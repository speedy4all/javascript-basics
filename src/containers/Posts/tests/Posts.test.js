import { render, screen, fireEvent } from '@testing-library/react'
import Posts from '../index';
import { PostProvider } from '../../../contexts/PostsContext'
import renderer from 'react-test-renderer';

describe('Posts component', () => {
    it('should match Snapshot', () => {
        const listPosts = renderer.create(<PostProvider><Posts /></PostProvider>).toJSON()
        expect(listPosts).toMatchSnapshot();
    })

    it('should clear searched value', () => {
        const { debug } = render(<PostProvider><Posts /></PostProvider>)
        debug();
        const inputSearch = screen.getByTestId('searchView-input');
        fireEvent.change(inputSearch, { target: { value: 'posts'}})
        expect(inputSearch.value).toBe('posts')
        const button = screen.getByTestId('clear-button')
        fireEvent.click(button);
        expect(inputSearch.value).toBe('')
    });
})
