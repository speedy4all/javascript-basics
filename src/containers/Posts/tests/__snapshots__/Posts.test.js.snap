// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component should match Snapshot 1`] = `
Array [
  <div
    data-testid="searchView-element"
  >
    <label
      htmlFor="searchInput"
    >
      Search post
    </label>
    <input
      data-testid="searchView-input"
      id="searchInput"
      onChange={[Function]}
      value=""
    />
    <button
      data-testid="clear-button"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>,
  <div
    style={
      Object {
        "padding": 16,
      }
    }
  />,
  <hr />,
  <div
    style={
      Object {
        "padding": 16,
      }
    }
  />,
]
`;
